logger/syslogish/server.go

```diff
@@ -0,0 +1,170 @@
+package syslogish
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/storage"
+)
+
+var appRegex *regexp.Regexp
+
+func init() {
+	appRegex = regexp.MustCompile(`^.* ([-_a-z0-9]+)\[[a-z0-9-_\.]+\].*`)
+}
+
+// Server implements a UDP-based "syslog-like" server.  Like syslog, as described by RFC 3164, it
+// expects that each packet contains a single log message and that, conversely, log messages are
+// encapsulated in their entirety by a single packet, however, no attempt is made to parse the
+// messages received or validate that they conform to the specification.
+type Server struct {
+	conn           net.PacketConn
+	listening      bool
+	queue          chan string
+	storageAdapter storage.Adapter
+	drain          drain.LogDrain
+	adapterMutex   sync.RWMutex
+	drainMutex     sync.RWMutex
+}
+
+// NewServer returns a pointer to a new Server instance.
+func NewServer(bindHost string, bindPort int) (*Server, error) {
+	addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", bindHost, bindPort))
+	if err != nil {
+		return nil, err
+	}
+	c, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &Server{conn: c, queue: make(chan string, 5)}, nil
+}
+
+// SetStorageAdapter permits a server's underlying storage.Adapter to be reconfigured (replaced)
+// at runtime.
+func (s *Server) SetStorageAdapter(storageAdapter storage.Adapter) {
+	// Get an exclusive lock before updating the internal pointer to the storage adapter.  Other
+	// goroutines holding read locks might depend on that pointer as it currently exists.
+	s.adapterMutex.Lock()
+	defer s.adapterMutex.Unlock()
+	s.storageAdapter = storageAdapter
+}
+
+// SetDrain permits a server's underlying drain.LogDrain to be reconfigured (replaced) at runtime.
+func (s *Server) SetDrain(drain drain.LogDrain) {
+	// Get an exclusive lock before updating the internal pointer to the log drain.  Other
+	// goroutines holding read locks might depend on that pointer as it currently exists.
+	s.drainMutex.Lock()
+	defer s.drainMutex.Unlock()
+	s.drain = drain
+}
+
+// Listen starts the server's main loop.
+func (s *Server) Listen() {
+	// Should only ever be called once
+	if !s.listening {
+		s.listening = true
+		go s.receive()
+		go s.process()
+		log.Println("syslogish server running")
+	}
+}
+
+func (s *Server) receive() {
+	// Make buffer the same size as the max for a UDP packet
+	buf := make([]byte, 65535)
+	for {
+		n, _, err := s.conn.ReadFrom(buf)
+		if err != nil {
+			log.Fatal("syslogish server read error", err)
+		}
+		message := strings.TrimSuffix(string(buf[:n]), "\n")
+		select {
+		case s.queue <- message:
+		default:
+		}
+	}
+}
+
+func (s *Server) process() {
+	for message := range s.queue {
```

Comment:
This is going to cause backpressure. It reads one message at a time off of the queue, and then locks at least three mutex objects. Depending on the backend, it may block until the receive can commit a write. This means that writes will be very slow, and the buffer will fill quickly. Then your sending process will begin to block, which will hold up your socket. In the worst cases, you'll start dropping UDP packets when they expire before they can be queued. If you are backed to something like an EBS volume, this sort of condition could obtain pretty easily because commits to disk are slow.

Strategies for avoiding backpressure are:

- increasing channel buffer size
- writing to an intermediate in-memory cache (ideally, something like a linked list)
- handling multiple messages during one lock/unlock cycle
  - x := len(queue)
  - lock
  - dequeue x items
  - unlock
  - repeat